### ISSUE TITLE
Problem with alignment of nanopore data using minimap2

### ISSUE THREAD

**Issue by SaraValsoni** [FelixKrueger/Bismark#722]
Hi Felix,

I have Nanopore reads from bisulfite converted amplicons and I would like to run Bismark (v0.24.1) to align them with Minimap2 (v2.28-r1209).
I'm using fastq files as input which were basecalled and demultiplexed by Dorado during the sequencing run. I merged the several fastq files produced for each sample using the cat command. Fastq files are gzipped.

I tried the following line of code:

`bismark --unmapped --ambiguous --ambig_bam -p 4 --non_directional -o /path/to/alignments_folder --path_to_minimap2 /path/to/minimap2/ --minimap2 --mm2_nanopore --mm2_maximum_length 100000 --genome bismark_minimap2_index -se barcode01_pass.fastq.gz &> barcode01_bismark.log`

The job starts but then it interrupts without any error. Both the unmapped and the ambiguous fastq/bam files are empty.

Here are the last lines in the log file:

    Ambiguous BAM output: barcode01_pass_bismark_mm2.ambig.bam

    >>> Writing bisulfite mapping results to barcode01_pass_bismark_mm2.bam <<<

    Unmapped sequences will be written to barcode01_pass.fastq.gz_unmapped_reads.fq.gz
    Ambiguously mapping sequences will be written to barcode01_pass.fastq.gz_ambiguous_reads.fq.gz

    Reading in the sequence file barcode01_pass.fastq.gz

And this is the content of the SE_report.txt file:

    Bismark report for: barcode01_pass.fastq.gz (version: v0.24.1)
    Option '--non_directional' specified: alignments to all strands were being performed (OT, OB, CTOT, CTOB)
    Bismark was run with minimap2 against the bisulfite genome of bismark_minimap2_index with the specified options: -a --MD -- 
    secondary=no -t 2 -x map-ont -K 250K

I tried running Minimap2 with the parameters specified by Bismark:

`/path/to/minimap2 -a --MD --secondary=no -t 2 -x map-ont -K 250K Bisulfite_Genome/CT_conversion/genome_mfa.CT_conversion.fa barcode01_pass.fastq.gz_C_to_T.fastq > minimap2_barcode01_CtoT.sam`

using the C_to_T converted reads (produced by the previous Bismark run) against the C_to_T converted genome (produced by bismark_genome_preparation) and it worked (actually it is still running).

Any idea to fix this issue?

Thank you!
Sara

**Comment by FelixKrueger** [comment]
Hi Sara,

The command you used looks a bit involved, but it doesn't actually seem to fail as such. How about you simpolify it somewhat and use a small test case to see if it works at all? I would suggest something like this:

```
bismark -u 1000 --non_directional -o /path/to/alignments_folder --path_to_minimap2 /path/to/minimap2/ --mm2  --genome bismark_minimap2_index barcode01_pass.fastq.gz 
```

This shouldn't take all that long....

**Comment by SaraValsoni** [comment]
Hi Felix,

thank you for your reply.
I tried running the code you suggested but I have the same result. The bam file is empty and the log/report files seems truncated.
Attached is the entire content of the log and report files in case they can be of help.

Is there anything else I can try?
Sara
[barcode01_bismark.log](https://github.com/user-attachments/files/18201027/barcode01_bismark.log)
[barcode01_pass_bismark_mm2_SE_report.txt](https://github.com/user-attachments/files/18201028/barcode01_pass_bismark_mm2_SE_report.txt)

**Comment by FelixKrueger** [comment]
The command I linked above does not redirect STDOUT or STDERR to a log file, but instead follows what Bismark is doing. 

To me it looks like the program is still running, and you just need to give it some more time. Can you check in a second terminal with `top` if CPU and memory are still being used? For the record, the Mapping report is initialised as you see it, and will be completed once the run is complete. When I did some benchmarking with minmap2 I found that some parameter choices make it really slow (even though you should be using the ones that came out if the test as good middle ground), see more here https://github.com/FelixKrueger/Bismark/issues/446.

**Comment by SaraValsoni** [comment]
Hi Felix,

Thank you for the link to the benchmarking you did with minimap 2 and sorry, I redirected the log file for my convenience.
I made an attempt checking the memory usage with `top` on a second terminal.
I can see a perl and 1 or 2 minimap2 jobs running then all of them disappear from the `top` output when on the other terminal a see the last line of the log file I sent to you:

`Reading in the sequence file ../barcode01_pass.fastq.gz`

Here the Bismark job stops (it is let's say released from the command line) and the memory usage drops.
After that I don't see any job still running on the `top` screen.

Do you think I should refer to the system administrators?
Thank you again for the help,
Sara

**Comment by FelixKrueger** [comment]
Once it reaches the stage `Reading in the sequence file ../barcode01_pass.fastq.gz` it should start comparing all four threads of minimap2...  I suspect that minimap2 for some reason doesn't report any alignments, and is working on large chunks that take a long time?  But because of this, the load of Bismark goes down (it will still be running though). Is there something that can be done about updating versions of minimap2, samtools (and possibly Bismark)?

**Comment by SaraValsoni** [comment]
Hi Felix,

We already had the last version of Minimap2, now we updated both Bismark and samtools.
Unfortunately the result is the same, the job starts then it stops at the same point and the bam file is empty.
I extracted the first 100 reads from my fastq file, could please have a look at them to be sure the file format is correct?

[barcode01_pass_100reads.fastq.gz](https://github.com/user-attachments/files/18213209/barcode01_pass_100reads.fastq.gz)

Thank you in advance,
Sara

**Comment by FelixKrueger** [comment]
Ha, I found it! Even though admittedly I done't know exactly what the problem really is...

Using your data, I ran pretty much what we discussed above. It starts all four alingment threads, and successfully finds the first alignments (as was printed onto screen/logfile. However as soon as all four threads are up and running, all 4 instances of minimap2 mysteriously day without further warning, leaving zombie processes:

```
[minimap2] <defunct>
[minimap2] <defunct>
[minimap2] <defunct>
[minimap2] <defunct>
```

I really don't know the reason for this, but I vaguely recalled that this could be an issue writing to BAM files (there seem to be some limitations with the BAM format specification....). When I chose SAM as output (`--sam`), it worked like a charm:

```
Final Alignment report
======================
Sequences analysed in total:    100
Number of alignments with a unique best hit from the different alignments:      100
Mapping efficiency:     100.0%

Sequences with no alignments under any condition:       0
Sequences did not map uniquely: 0
Sequences which were discarded because genomic sequence could not be extracted: 0

Number of sequences with unique best (first) alignment came from the bowtie output:
CT/CT:  45      ((converted) top strand)
CT/GA:  1       ((converted) bottom strand)
GA/CT:  53      (complementary to (converted) top strand)
GA/GA:  1       (complementary to (converted) bottom strand)

Final Cytosine Methylation Report
=================================
Total number of C's analysed:   24776

Total methylated C's in CpG context:    2406
Total methylated C's in CHG context:    148
Total methylated C's in CHH context:    233
Total methylated C's in Unknown context:        10

Total unmethylated C's in CpG context:  2623
Total unmethylated C's in CHG context:  7810
Total unmethylated C's in CHH context:  11556
Total unmethylated C's in Unknown context:      50

C methylated in CpG context:    47.8%
C methylated in CHG context:    1.9%
C methylated in CHH context:    2.0%
C methylated in Unknown context (CN or CHN):    16.7%


Bismark completed in 0d 0h 1m 11s

====================
Bismark run complete
```

I hope that this may give you additional pointers to search the internet and/or the minimap2 forum for? (as an aside `--cram also fails, similar to the default BAM format).

**Comment by SaraValsoni** [comment]
Hi Felix,

Great! Finally I found the problem! Samtools is not able to convert the sam into bam format because of the long read name:

`[E::sam_parse1] query name too long`
`[W::sam_read1_sam] Parse error at line 28`
`samtools view: error reading file "barcode01_pass.sam"`

I found that someone else already fixed this issue by reducing the read name. I tried the same and samtools works!
So now I am running bismark using the fastq files with a shorter read name and it is working also without the `--sam` parameter.

Thank you for the help!
Sara
